include/node.hpp

```diff
@@ -306,27 +301,57 @@ class node : protected std::tuple<Arguments...> {
 
     work_return_t
     work() noexcept {
-        // Capturing structured bindings does not work in Clang...
-        const auto inputs_status = self().inputs_status(self());
+        using input_types = traits::node::input_port_types<Derived>;
+        using output_types = traits::node::output_port_types<Derived>;
 
-        if (inputs_status.available_values_count == 0) {
-            return inputs_status.at_least_one_input_has_data ? work_return_t::INSUFFICIENT_INPUT_ITEMS : work_return_t::DONE;
+        constexpr bool is_source_node = input_types::size == 0;
+        constexpr bool is_sink_node = output_types::size == 0;
+
+        std::size_t samples_to_process = 0;
+        if constexpr (is_source_node) {
+            if constexpr (requires(const Derived &d) {
+                              { available_samples(d) } -> std::same_as<std::size_t>;
+                          }) {
+                // the (source) node wants to determine the number of samples to process
+                samples_to_process = available_samples(self());
+            } else if constexpr (is_sink_node) {
+                // no input or output buffers, derive from internal "buffer sizes" (i.e. what the
+                // buffer size would be if the node were not merged)
+                constexpr std::size_t chunk_size = Derived::merged_work_chunk_size();
+                static_assert(
+                        chunk_size != std::dynamic_extent && chunk_size > 0,
+                        "At least one internal port must define a maximum number of samples or the non-member/hidden "
+                        "friend function `available_samples(const NodeType&)` must be defined.");
+                samples_to_process = chunk_size;
+            } else {
+                // derive value from output buffer size
+                samples_to_process
+                        = std::apply([&](const auto &...ports) { return std::min({ ports.writer().available()... }); },
```

Comment:
I like this approach ... the limit should also take into account the MIN/MAX limits that the user imposed onto the ports.

Reply:
true, will change

Reply:
Questions:
1. Should `min_buffer_size()` only be checked for the `is_source_node` case?
2. Should it be checked if the `available_samples` customization point is implemented? If yes, how will the "EOF" case, where only a few samples are left, work?
4. If it's also checked for not-source-nodes, and the number of samples is determined by `inputs_status`, what's the return code on failure? `work_return_t::INSUFFICIENT_INPUT_ITEMS`, right? Whereas on a source node it's `work_return_t::INSUFFICIENT_OUTPUT_ITEMS`?

Reply:
Re 2:

I think the existing/proposed solution of `if constexpr (requires(const Derived &d) {..} ` plus the other to if-else cases is IMO good as is.